File: Readio/src/main/java/ninegle/Readio/book/domain/Review.java
package ninegle.Readio.book.domain;

import java.math.BigDecimal;
import java.time.LocalDateTime;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import ninegle.Readio.user.domain.User;

/**
 * Readio - Review
 * create date:    25. 5. 9.
 * last update:    25. 5. 9.
 * author:  gigol
 * purpose: 
 */

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)

@Table(name = "review")
public class Review {
	@Builder(toBuilder = true)
	public Review(BigDecimal rating, String text, User user, Book book) {
		this.rating = rating;
		this.text = text;
		this.user = user;
		this.book = book;
	}

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	private BigDecimal rating;

	@Column(columnDefinition = "TEXT")
	private String text;

	private LocalDateTime createdAt;

	private LocalDateTime updatedAt;

	@ManyToOne
	@JoinColumn(name = "user_id", nullable = false)
	private User user;

	@ManyToOne
	@JoinColumn(name = "book_id", nullable = false)
	private Book book;

	@PrePersist
	public void onCreate() {
		this.createdAt = LocalDateTime.now();
		this.updatedAt = this.createdAt;
	}

	@PreUpdate
	public void onUpdate() {
		this.updatedAt = LocalDateTime.now();
	}
}

File: Readio/src/main/java/ninegle/Readio/book/domain/Review.kt
package ninegle.Readio.book.domain

import jakarta.persistence.*
import ninegle.Readio.book.dto.reviewdto.ReviewRequestDto
import ninegle.Readio.user.domain.User
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
class Review (

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    var rating: BigDecimal,

    @field:Column(
        columnDefinition = "TEXT"
    )
    var text: String,

    @field:JoinColumn(
        name = "user_id",
        nullable = false
    )
    @field:ManyToOne
    var user: User,

    @field:JoinColumn(
        name = "book_id",
        nullable = false
    )
    @field:ManyToOne
    var book: Book,

    var createdAt: LocalDateTime? = null,
    var updatedAt: LocalDateTime? = null
) {

    @PrePersist
    fun onCreate() {
        this.createdAt = LocalDateTime.now()
        this.updatedAt = this.createdAt
    }

    @PreUpdate
    fun onUpdate() {
        this.updatedAt = LocalDateTime.now()
    }

    fun update(dto: ReviewRequestDto) : Review{
        this.rating = dto.rating
        this.text = dto.text
        return this
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Review

        if (id != other.id) return false
        if (rating != other.rating) return false
        if (text != other.text) return false
        if (user != other.user) return false
        if (book != other.book) return false
        if (createdAt != other.createdAt) return false
        if (updatedAt != other.updatedAt) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id?.hashCode() ?: 0
        result = 31 * result + rating.hashCode()
        result = 31 * result + text.hashCode()
        result = 31 * result + user.hashCode()
        result = 31 * result + book.hashCode()
        result = 31 * result + (createdAt?.hashCode() ?: 0)
        result = 31 * result + (updatedAt?.hashCode() ?: 0)
        return result
    }
}
